[Flow Aggregator] Update elk flow collector for flow aggregator

zyiou opened the pull request:
fixes #1707, fixes #1794

Logstash changes:
- add new fields for flow aggregator in ipfix definitions file
- modify `ipfix.bytes` and `ipfix.packets` from `octetDeltaCount` to `octetTotalCount`
- add mutate functions to create flow key 
   - Pod-to-Pod flow: sourcePodName:port, destinationPodName:port, protocol
   - Pod-to-Service flow: sourcePodName:port, destinationServiceName:port, protocol
- add interval as a parameter for calculating Tx and Rx in Kibana dashboard
- integrate filters into a ruby file, support dynamic interval (flowEndSeconds-last export time) for calculating throughput

Dashboard changes:
- make pod-to-service flow as individual dashboard and add Tx and Rx by Flows graphs
- add heatmap for node throughput
- add Tx and Rx by Flows graphs to pod-to-pod flow dashboard
- add flow key filters for overview, pod-to-pod flow, pod-to-service-flow and flow records dashboards
- change all stacked line graph to line graph, making it easier for users to get the exact value
- remove "Antrea:" prefix in all visualization names
- change all the x-axis of time series data from timestamp of receiving to `ipfix.flowEndSeconds` (actual flow export time)

Corresponding images in documentation will be uploaded when PR gets approved.

Overview dashboard:
![overview](https://user-images.githubusercontent.com/36459546/106104185-f135fa80-60f6-11eb-82a8-6a0b09946d54.png)


Pod-to-Pod Flow dashboard:
![pod-to-pod 1](https://user-images.githubusercontent.com/36459546/106104203-f72bdb80-60f6-11eb-987a-35b8e2989273.png)
![pod-to-pod 2](https://user-images.githubusercontent.com/36459546/106507252-a2ce8600-647f-11eb-9bb7-f4ab8cfd9e80.png)
![pod-to-pod 3](https://user-images.githubusercontent.com/36459546/106507256-a3671c80-647f-11eb-82fe-c6d340c42cc5.png)



Pod-to-Service Flow dashboard:
![pod-to-service 1](https://user-images.githubusercontent.com/36459546/106104226-027f0700-60f7-11eb-8dee-cb1764169134.png)
![pod-to-service 2](https://user-images.githubusercontent.com/36459546/106507241-9cd8a500-647f-11eb-8881-499e99ef216b.png)
![pod-to-service 3](https://user-images.githubusercontent.com/36459546/106507248-a104c280-647f-11eb-8148-942e2493dba3.png)


Node dashboard:
![node 1](https://user-images.githubusercontent.com/36459546/106104309-25112000-60f7-11eb-959e-04031fdedf44.png)
![node 2](https://user-images.githubusercontent.com/36459546/106998840-fc61d980-6739-11eb-8713-54dba4bd371e.png)


Flow Records dashboard:
![flow-record 1](https://user-images.githubusercontent.com/36459546/106507297-afeb7500-647f-11eb-892c-89cfb3200968.png)

jianjuns reviewed (commented):
I just looked at the screenshots. Two comments:
1. The ordering of source/destination Namespace, Name, Node fields are not correct in some views. Probably we should always follow the same order, put Name after Namepace, and put fields of source or destination together (and source before destination).
2. Why for Pod-Service flows, the Service Namespace field can not be selected?

zyiou commented:
> I just looked at the screenshots. Two comments:
> 
> 1. The ordering of source/destination Namespace, Name, Node fields are not correct in some views. Probably we should always follow the same order, put Name after Namepace, and put fields of source or destination together (and source before destination).

You mean the filters, right?
> 2. Why for Pod-Service flows, the Service Namespace field can not be selected?

Thanks for reminding. will add the Service Namespace filter.

jianjuns commented:
Yes, I mean filters.

srikartati reviewed (commented):
From dashboard snapshots in description:
- Could the snippet "All stats are from..." be moved to the end of the page in dashboards?
- In node dashboard, I think a [heatmap chart](https://www.elastic.co/blog/awesome-new-kibana-visualizations-heatmap-and-point-series) for Aggregated total Ingress throughput (Rx Mbps) and Aggregated total egress throughput (Tx traffic) can be useful to quickly identify which node takes a lion share of bandwidth.

zyiou commented on build/yamls/elk-flow-collector/logstash/logstash.conf:
```diff
@@ -28,7 +28,7 @@ filter {
     } else {
         mutate {
-            replace => { "[ipfix][bytes]" => "%{[ipfix][octetDeltaCount]}" }
+            replace => { "[ipfix][bytes]" => "%{[ipfix][reverseOctetDeltaCount]}" }
```
I don't think so. The issue is related to packetDeltaCount rather than octetDeltaCount. It will be better if an issue is created to describe the problem in detail.

srikartati replied:
@zyiou Not sure Piotr will open an issue. I think there is enough context there to fix the dashboard. It is good to improve the dashboard through feedback from Slack in addition to Issues on github repo. 

What I understood from the thread: there is an issue if there is only one flow record for the flow (short flows), then we do not show any data for those flows. 
Seems like ipfix.Packets and ipfix.Bytes are the delta counts. Could we look at changing to total packet counts

antoninbas replied:
@zyiou please feel free to open the issue yourself with the information from the Slack thread so we can track this

zyiou replied:
@srikartati I read the thread again. It seems that the only remaining problem to be solved is changing ipfix.packets and ipfix.bytes to total packet and byte counts. Created an issue [here](https://github.com/vmware-tanzu/antrea/issues/1794). Let me know if I missed something. Thanks!

zyiou commented on build/yamls/elk-flow-collector/logstash/logstash.conf:
```diff
@@ -61,7 +89,12 @@ filter {
         '
     }
     mutate {
+        add_field => {"[ipfix][interval]" => 60 }
```
Yes. will add the info in doc. Thanks!

srikartati commented on docs/network-flow-visibility.md:
````diff
@@ -367,7 +367,11 @@ git clone --depth 1 --branch main https://github.com/vmware-tanzu/antrea.git &&
+To configure the export interval as `flowExportInterval` in [Configuration](#configuration), 
+modify following `interval` value (in seconds) in `elk-flow-collector/logstash/logstash.conf`
+```
+script_params => { "interval" => 60 }
````
With dynamic interval tracking for flow records, is this still valid?

zyiou replied:
It is valid for the first record.

srikartati reviewed (commented):
Dashboard images look great. Thanks for fixing the ipfix.packets and ipfix.bytes issue.

Couple of things:
- In the node heatmap, is it expected to have flowEndSeconds in x-axis? What does it signify?
- Please look at the toc error after making changes in the doc.

jianjuns commented on docs/network-flow-visibility.md:
````diff
@@ -367,7 +367,11 @@ git clone --depth 1 --branch main https://github.com/vmware-tanzu/antrea.git &&
 svn export https://github.com/vmware-tanzu/antrea/trunk/build/yamls/elk-flow-collector/
 ```
-
+To configure the export interval as `flowExportInterval` in [Configuration](#configuration), 
````
Heard from @antoninbas: we need a new line before and after a code block or a bulleted list for antrea.io to show the document correctly.

antoninbas replied:
that's correct, but we have a CI job for that now so I don't hav to point it out anymore :)

jianjuns commented on docs/network-flow-visibility.md:
````diff
@@ -367,7 +367,11 @@ git clone --depth 1 --branch main https://github.com/vmware-tanzu/antrea.git &&
-
+To configure the export interval as `flowExportInterval` in [Configuration](#configuration), 
+modify following `interval` value (in seconds) in `elk-flow-collector/logstash/logstash.conf`
+```
````
Ditto.

zyiou replied:
Fixed. Thanks!

jianjuns commented on docs/network-flow-visibility.md:
````diff
@@ -367,7 +367,11 @@ git clone --depth 1 --branch main https://github.com/vmware-tanzu/antrea.git &&
 ```
-
+To configure the export interval as `flowExportInterval` in [Configuration](#configuration), 
+modify following `interval` value (in seconds) in `elk-flow-collector/logstash/logstash.conf`
````
"the following" or just "the".

jianjuns commented on docs/network-flow-visibility.md:
```diff
@@ -397,37 +401,41 @@ An overview of Pod-based flow records information is provided.
 Visualization Flows Dashboard">
 
-<img src="https://downloads.antrea.io/static/flow-visualization-flow-4.png" width="900" alt="Flow
+Aggregated Tx and Rx traffic based on destination Service are shown in line graph.
```
are -> is

jianjuns reviewed (commented):
The dashboards looks cool!

The words of "All stats are from soure Node..." will be shown in the real dashboard? Are they necessary? Do not look pretty.

Do you think it looks better to put "Destination Pod Namespace" in the same line as other destination Pod filters? How about move "Flow Keys" to the source Pod line if needed?

zyiou commented:
> Dashboard images look great. Thanks for fixing the ipfix.packets and ipfix.bytes issue.
> 
> Couple of things:
> 
> * In the node heatmap, is it expected to have flowEndSeconds in x-axis? What does it signify?

Kibana uses the default label there. Changed the label to Timestamp. But the timestamp is actually using the value of flowEndSeconds.

> * Please look at the toc error after making changes in the doc.

Thanks for reminding. Updated.

zyiou commented:
> The dashboards looks cool!
> 
> The words of "All stats are from soure Node..." will be shown in the real dashboard? Are they necessary? Do not look pretty.
> 

yes it will be shown. We have another set of mirrored graphs to show Tx and Rx from the data at destination node which are theoretically duplications. To avoid displaying both set of graphs, we add this annotations. Maybe we can move the annotation to our doc instead of showing it in the dashboard?

> Do you think it looks better to put "Destination Pod Namespace" in the same line as other destination Pod filters? How about move "Flow Keys" to the source Pod line if needed?

Actually arrangement of the filters is responsive to browser width. If we narrow the browser width, Destination Pod Namespace will be shown in the next line like:
<img width="1262" alt="Screen Shot 2021-01-29 at 5 17 08 PM" src="https://user-images.githubusercontent.com/36459546/106342940-cc0ccd80-6257-11eb-9f5f-22939908c382.png">

jianjuns commented:
> yes it will be shown. We have another set of mirrored graphs to show Tx and Rx from the data at destination node which are theoretically duplications. To avoid displaying both set of graphs, we add this annotations. Maybe we can move the annotation to our doc instead of showing it in the dashboard?

I think we either remove the annotation, or at least use small size font (the current not looks good). 

> Actually arrangement of the filters is responsive to browser width.
Got it!

srikartati commented:
> > Dashboard images look great. Thanks for fixing the ipfix.packets and ipfix.bytes issue.
> > Couple of things:
> > 
> > * In the node heatmap, is it expected to have flowEndSeconds in x-axis? What does it signify?
> 
> Kibana uses the default label there. Changed the label to Timestamp. But the timestamp is actually using the value of flowEndSeconds.

Ohh. Why does the timestamp have to be flowEndSeconds? Node heat map can be determined by aggregating flows and computing Rx/TxMbps per 60seconds like in other cases, right?

zyiou commented:
> > > Dashboard images look great. Thanks for fixing the ipfix.packets and ipfix.bytes issue.
> > > Couple of things:
> > > 
> > > * In the node heatmap, is it expected to have flowEndSeconds in x-axis? What does it signify?
> > 
> > 
> > Kibana uses the default label there. Changed the label to Timestamp. But the timestamp is actually using the value of flowEndSeconds.
> 
> Ohh. Why does the timestamp have to be flowEndSeconds? Node heat map can be determined by aggregating flows and computing Rx/TxMbps per 60seconds like in other cases, right?

Now in other cases we all use flowEndSeconds as timestamp (suggested by @dreamtalen) since it reflects the actual time when flow gets exported rather than the time elk flow collector received it. We thought flowEndSeconds is more meaningful and close to flow happening time. When calculating Rx/Tx Mbps, we aggregate throughput as a unit of 60 seconds also based on flowEndSeconds. Do you think it would be better to use default timestamp instead?

srikartati commented:
> > > > Dashboard images look great. Thanks for fixing the ipfix.packets and ipfix.bytes issue.
> > > > Couple of things:
> > > > 
> > > > * In the node heatmap, is it expected to have flowEndSeconds in x-axis? What does it signify?
> > > 
> > > 
> > > Kibana uses the default label there. Changed the label to Timestamp. But the timestamp is actually using the value of flowEndSeconds.
> > 
> > 
> > Ohh. Why does the timestamp have to be flowEndSeconds? Node heat map can be determined by aggregating flows and computing Rx/TxMbps per 60seconds like in other cases, right?
> 
> Now in other cases we all use flowEndSeconds as timestamp (suggested by @dreamtalen) since it reflects the actual time when flow gets exported rather than the time elk flow collector received it. We thought flowEndSeconds is more meaningful and close to flow happening time. When calculating Rx/Tx Mbps, we aggregate throughput as a unit of 60 seconds also based on flowEndSeconds. Do you think it would be better to use default timestamp instead?

FlowEndSeconds belong to just one flow and not defined for multiple flows. When interpreting flow time-series we might be using FlowEndSeconds to plot the X-axis and ultimately use flowEndSeconds of different flows to get the aggregated throughput at that specific time instance. Representing X-axis as generic time is less confusing than using flowEndSeconds. The latter could be interpreted in different ways.

srikartati reviewed (dismissed):
LGTM.
Thanks for addressing the comments.

srikartati commented:
@jianjuns I see that Yiou reduced the font size. Hope that is ok for you. If there are no other comments from you, I will merge this PR.

srikartati commented:
These changes do not affect the Antrea tests, so skipping them.
/skip-all

srikartati reviewed (approved):
Approving again.
